### PR TITLE
Jetpack Social: Add sidebar panel for Social Image Generator

### PR DESCRIPTION
**Pull request by danielpost**
This change copies over the existing SIG panel into the Jetpack Social panel. AND also sets up the post-meta usage.

## Proposed changes:

* Added new panel
* Created a hook which is used to edit and read the post meta for the SIG options.
* Templates are not part of this change.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

1203750066198187-as-1203857426851217

## Does this pull request change what data or activity we track or use?

no.

## Testing instructions:

* Go to the Social editor tab, you should not see the SIG panel there.
* Add the `publicize-social-image-generator` blog sticker, then go to the Social editor tab, you should see a new panel there called Social Image Generator
* Disabling/Enabling should work. Custom text can be set and saved. Image type is saved.
* Select custom image. You should see the same image picker as with Social, and it should work and save to meta.
* ~~Until #28773 is not merge, you need to rebase this on that to be able to test it.~~

### CODE REVIEW

File: projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js
Comment (pablinos) [inline projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js]: I'm not sure, but it could possibly be slightly more efficient to split this into two calls to `useSelect`, one for each store p7H4VZ-2Q6-p2#comment-9139

File: projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js
Comment (pablinos) [inline projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js]: Should these be in `useCallbacks` too? I think `editPost` should always be stable, so if we're memoizing on `currentOptions` for one than perhaps we should for the others, or remove the `useCallback` for `updateSettings`

File: projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js
Comment (pablinos) [inline projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js]: This might well be in the realms "clever"/unreadable code, but something like this might prevent the multiple calls to (and the need for) `getSetting` or perhaps we could change that to `getSettings` which could return the object with the properties we need?

```suggestion
		...( ( {
		    enabled: isEnabled = ! isPostPublished,
		    custom_text: customText = null,
		    image_type: imageType = null,
		    image_id: imageId = null
		} ) => ( {	    
		    isEnabled,
		    customText,
		    imageType,
		    imageId,
		} )( currentOptions?.image_generator_settings ?? {} )
```

File: projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js
Comment (pablinos) [inline projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js]: Related to my comment above `updateSettings` is only used in `updateSetting` so perhaps we could combine the functions in the `useCallback`

File: projects/js-packages/publicize-components/src/store/selectors.js
Comment (pablinos) [inline projects/js-packages/publicize-components/src/store/selectors.js]: I've just realised that we're using lodash here and I think we're trying to move away from it. Perhaps we could start with this? Do something like:

```suggestion
	return getJetpackSocialOptions()?.[ 'image_generator_settings' ] ?? [];
```

File: projects/plugins/social/src/js/editor.js
Comment (pablinos) [inline projects/plugins/social/src/js/editor.js]: Is there a feature check here that will prevent this from rendering without the plan/feature flag?

**Review by pablinos — COMMENTED** [review]
Looks great to me. I just had a few minor thoughts, and I don't think we can merge this until we have the feature flag in place - or at least we need to get it in place before the next release, but it would be safer not to merge this first.

I think this is something for a follow-up but could we make the media picker have support for the alternative sources that Jetpack provides, like on the featured image and image blocks?

image

File: projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js
Comment (gmjuhasz) [reply projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js]: Done!

File: projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js
Comment (gmjuhasz) [reply projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js]: Yeah good point, I merged the two.

File: projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js
Comment (gmjuhasz) [reply projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js]: I tried to clean it up a bit, what do you think of this?

File: projects/plugins/social/src/js/editor.js
Comment (gmjuhasz) [reply projects/plugins/social/src/js/editor.js]: Yeah, I'll add it after we merge it here: #29001

File: projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js
Comment (pablinos) [inline projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js]: Real nitpick, but I think this could be:
```suggestion
	const getCurrentSettings = sigSettings => ( {
		isEnabled: sigSettings?.enabled ?? ! isPostPublished,
		customText: sigSettings?.custom_text ?? null,
		imageType: sigSettings?.image_type ?? null,
		imageId: sigSettings?.image_id ?? null,
	) };
```

File: projects/plugins/social/src/js/editor.js
Comment (pablinos) [reply projects/plugins/social/src/js/editor.js]: Nice!

File: projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js
Comment (pablinos) [reply projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js]: Thanks! I'm not sure if it's beneficial or not, but just in case :)

**Review by pablinos — COMMENTED** [review]
Great, this looks good. I'll hold off approving it for now until we incorporate the feature check in #29001

We can make the changelogger happy at the same time.

File: projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js
Comment (pablinos) [inline projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js]: I've just realised that we could define this function outside of the hook, but maybe that's an optimisation too far!

**Review by pablinos — COMMENTED** [review]
Nice! I won't approve it yet as we're waiting on #29001, but I reckon this will be good to merge once we add the feature check.

File: projects/plugins/social/src/js/editor.js
Comment (pablinos) [inline projects/plugins/social/src/js/editor.js:130]: I've only just realised that this adds it to Social, but not the core Jetpack plugin. I'm not sure if we would do that [here](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/extensions/plugins/publicize/index.js#L49) or [as a component here](https://github.com/Automattic/jetpack/blob/b313bd253afdf919adf7a71a866d0ab85b3a7bda/projects/plugins/jetpack/extensions/blocks/social-previews/index.js#L21)

**Review by pablinos — DISMISSED** [review]
Looks good. We can add it to Jetpack in a follow up, or add it to this one.

File: projects/js-packages/publicize-components/src/components/social-image-generator/panel/index.js
Comment (pablinos) [inline projects/js-packages/publicize-components/src/components/social-image-generator/panel/index.js]: This change makes the layout of the pre-publish side bar a bit odd. There's a gap at the bottom of the panel, but also this panel becomes nested in the 'Share this post' one.

CleanShot 2023-02-23 at 10 38 00@2x

image

**Review by pablinos — COMMENTED** [review]
If possible, I think we should try and preserve the behaviour of the pre-publish sidebar. If it's too tricky, maybe we could make it a section of the PublicizePanel, like the media settings.

File: projects/js-packages/publicize-components/src/components/social-image-generator/panel/index.js
Comment (gmjuhasz) [reply projects/js-packages/publicize-components/src/components/social-image-generator/panel/index.js]: Yeah good catch. Following our conversation on Slack, I restored the previous version, and added it to the Jetpack side as well. It is duplicated like this for now, but it will be removed with the next UI change anyway

Comment (pablinos) [reply projects/js-packages/publicize-components/src/components/social-image-generator/panel/index.js]: Yes, I think that makes sense for now. Thanks!

**Review by pablinos — APPROVED** [review]
LGTM!

File: projects/plugins/social/src/js/editor.js
Comment (dereksmart) [reply projects/plugins/social/src/js/editor.js:130]: Let's hold off on worrying about bringing the feature to the Jetpack plugin. Prioritize building and making it available in the Social Plugin.

**Pull request merged**
